### DISCUSSION TITLE
My code doesn't (fully) work

### DISCUSSION THREAD

**Discussion by rec32** [rejetto/hfs#818] [Q&A]
Hello rejetto, maybe you remember me, I'm sorry to bother you again, but, I'm completely stuck.
(also hello to whoever else reading this, not to be rude)

I am using following code to display custom file and folder icons: [1_works.txt](https://github.com/user-attachments/files/17891657/1_works.txt)
(I have to attach it since ` symbols break github formatting)

it works, but it's dumb, it doesn't check if pic is actually present in folder, also it only checks for first pic

so i tried this code, similar to those i was using in delphi version of hfs: [2_does_not_work.txt](https://github.com/user-attachments/files/17891658/2_does_not_work.txt)
it's much better on paper, however it doesn't work at all. i tried so many variants of it, always failing.

what am i doing wrong? guess i'm misusing "exists" function? or loop? or something else?

**Comment by rejetto** [comment]
hi!

Let's summarize what you are doing for other readers: you are deciding what file to use as icon for folders, based on what files are available on the server.

First: there's no such thing as the "exists" function *in your context*, and i'm sure you get an error on that.
If you are not sure about the error, you have to open the browser's console (F12).

Programming for client-server is not the easiest of things.
The code you are writing is clearly meant for the browser. On the browser-side you cannot do the same things as on the server-side.
The browser doesn't know if a file exists, as it runs on another machine. At best, the browser MAY try to look at the list sent by hfs when it's completed, and see if it's there, but this doens't even sound like a good idea, as you have to wait for the whole list (that can take long if you are searching in a deep big folder), or the browser can ASK the server about it (a "head" request, in http terms, without content), but it means that you make an extra request for each file. It would be ok on small folders on servers that don't suffer in performance, but not ideal in the general case. Generally, I don't suggest this.

The other approach to the problem is do all the work directly on the server (AKA back-end), that is "plugin.js".
Here you need "onDirEntry", and you can do the work fast. You can use "exists" after you "require" it from the "fs" module, as exists is not a native function of javascript. Most functions must be required.
If you are not knowledgable about async, i suggest you to use [existsSync](https://nodejs.org/docs/latest-v18.x/api/fs.html#fsexistssyncpath) instead.
Once you have your information (what icon to use), you add it inside the entry, like `entry.myIcon = someFile`, and it will be included in the entry the browser gets.
Now the browser will find this information, ready to use.
